backend/src/main/java/com/bakdata/conquery/models/auth/oidc/passwordflow/OIDCResourceOwnerPasswordCredentialRealm.java

```diff
@@ -0,0 +1,220 @@
+	 * We only hold validated Tokens for some minutes to recheck them regulary with Keycloak.
+	 */
+	private LoadingCache<JwtToken, TokenIntrospectionSuccessResponse> tokenCache = CacheBuilder.newBuilder()
+		.expireAfterWrite(10, TimeUnit.MINUTES)
+		.build(new TokenValidator());
+	
+	@Override
+	protected void onInit() {
+		super.onInit();
+		this.setCredentialsMatcher(new SkippingCredentialsMatcher());
+		this.setAuthenticationTokenClass(TOKEN_CLASS);
+		this.clientAuthentication = new ClientSecretBasic(new ClientID(config.getResource()), new Secret((String)config.getCredentials().get("secret")));
+		
+		authzClient = AuthzClient.create(config);
+		serverConf = authzClient.getServerConfiguration();
+	}
+	
+	@Override
+	@SneakyThrows
+	protected ConqueryAuthenticationInfo doGetConqueryAuthenticationInfo(AuthenticationToken token) throws AuthenticationException {
+		
+		TokenIntrospectionSuccessResponse successResponse = tokenCache.get((JwtToken) token);
+		
+		
+		if(isExpired(successResponse)){
+			tokenCache.invalidate(token);
+			throw new ExpiredCredentialsException();
+		}
+
+		String username = successResponse.getUsername();
+		if(StringUtils.isBlank(username)) {
+			username = successResponse.getStringParameter("preferred_username");
+		}
+		if(StringUtils.isBlank(username)) {
+			throw new IllegalStateException("Unable to retrieve a user identifier from validated token. Dismissing the token.");
+		}
+		
+		UserId userId = new UserId(username);
+		User user = storage.getUser(userId);
+		// try to construct a new User if none could be found in the storage
+		if (user == null) {
+			String userLabel = successResponse.getStringParameter("name");
+			user = new User(username, userLabel != null ?  userLabel : username);
+			storage.addUser(user);
+			log.info(
+				"Created new user: {}",
+				user);
+		}
+
+		return new ConqueryAuthenticationInfo(user.getId(), token, this, true);
+	}
+	
+	private boolean isExpired(TokenIntrospectionSuccessResponse tokenInstrospection) {
+		LocalDateTime expTime = tokenInstrospection.getExpirationTime().toInstant().atZone(ZoneId.systemDefault()).toLocalDateTime();
+		LocalDateTime now = LocalDateTime.now();
+		
+		boolean result = expTime.isBefore(now);
+		if(result) {
+			log.debug("Provided token expired at {} ( now is {})", expTime, now);
+		}
+		return result;
+	}
+
+	/**
+	 * Is called by the CacheLoader, so the Token is not validated on every request.
+	 */
+	private TokenIntrospectionSuccessResponse validateToken(AuthenticationToken token) throws ParseException, IOException {
+		TokenIntrospectionRequest request = new TokenIntrospectionRequest(URI.create(serverConf.getTokenIntrospectionEndpoint()) , clientAuthentication, new TypelessAccessToken((String) token.getCredentials()));
+				
+		TokenIntrospectionResponse response = TokenIntrospectionResponse.parse(request.toHTTPRequest().send());
+		
+		if (!response.indicatesSuccess()) {
+			log.error(response.toErrorResponse().getErrorObject().toString());
```

Comment:
Könntest du das ErrorObject nicht als Cause in die Exception stecken?

Reply:
Das ist leider kein Throwable